### PR TITLE
Fix bdalg.connect

### PR DESCRIPTION
**Pull request by sawyerbfuller**
Fixes bug in index check in `bdalg.connect` described in #421. 

Also added a few more checks, helpful error messages, and unit tests.

### CODE REVIEW

**Comment by roryyorke** [comment]
Looks good to me.  One minor stylistic comment: you don't need `as context` if you're not going to use `context`.

File: control/bdalg.py [hunk 1]
```diff
@@ -325,15 +327,34 @@ def connect(sys, Q, inputv, outputv):
...
+        raise IndexError( 
+            "outputv index %s out of bounds"%outputv[np.where(index_errors)])
+    index_errors = (Q[:,0:1] - 1 > sys.inputs) | (Q[:,0:1] < 1)
+    if np.any(index_errors):  
+        raise IndexError( 
+            "Q input index %s out of bounds"%Q[np.where(index_errors)])
+    index_errors = (np.abs(Q[:,1:]) - 1 > sys.outputs)
+    if np.any(index_errors):  
+        raise IndexError( 
+            "Q output index %s out of bounds"%Q[np.where(index_errors)])
```
Comment (bnavigator) [inline control/bdalg.py:347]: - Style issue: no white space between format strings and argument. 
- %s is not the correct format. Indices are integers.

File: control/tests/bdalg_test.py [hunk 2]
```diff
@@ -270,6 +272,46 @@ def test_feedback_args(self):
         sys = ctrl.feedback(1, frd)
         self.assertTrue(isinstance(sys, ctrl.FRD))
 
+    def testConnect(self):
```
Comment (bnavigator) [inline control/tests/bdalg_test.py:275]: These tests only look for failures. What about tests that successfully connect systems?

Obviously I will pytest.parametrize this for #438

**Review by bnavigator — CHANGES_REQUESTED** [review]
Looks good, thanks! One step closer to 0.8.4 :tada: 

Some more comments about style and the tests.

File: control/tests/bdalg_test.py (hunk 2 above)
Comment (sawyerbfuller) [reply control/tests/bdalg_test.py:275]: > These tests only look for failures. What about tests that successfully connect systems?

good idea; added. 
 
> Obviously I will pytest.parametrize this for #438

thx!

File: control/bdalg.py (hunk 1 above)
Comment (sawyerbfuller) [reply control/bdalg.py]: fixed style issue.

re print: it to prints out an array of failing indices. is there a better way than just casting to a string like this?

File: control/bdalg.py [hunk 3]
```diff
@@ -302,14 +302,16 @@ def connect(sys, Q, inputv, outputv):
...
-        Interconnection matrix. First column gives the input to be connected
-        second column gives the output to be fed into this input.  Negative
-        values for the second column mean the feedback is negative, 0 means
-        no connection is made.  Inputs and outputs are indexed starting at 1.
+        Interconnection matrix. First column gives the input to be connected.
+        The second column gives the index of an output that is to be fed into 
+        that input. Each additional column gives the index of an additional 
+        input that may be optionally added to that input. Negative
+        values mean the feedback is negative. A zero value is ignored. Inputs 
+        and outputs are indexed starting at 1 to communicate sign information.
```
[the remark this answers is not in the document]
Comment (sawyerbfuller) [reply control/bdalg.py]: found it i think

File: control/bdalg.py (hunk 1 above)
Comment (bnavigator) [reply control/bdalg.py]: You're right. According to the [doc](https://docs.python.org/3/library/stdtypes.html#old-string-formatting), `%s` in old style formatting converts any argument using `str()`, which should always work for an array of integers.

**Review by bnavigator — APPROVED** [review]

**Pull request merged**